### PR TITLE
Testing specs for Enumerable project

### PR DESCRIPTION
**Pull request by enelesmai**
* Setup Rspec for project
* Used existing repository of Enumerables
* Added specs for all customized Enumerables methods
* Wrapped specs for each method in describe block
* Using let to define a variable for the test cases
* Using it block for each expected behaviour

### CODE REVIEW

File: spec/enumerable_spec.rb [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+  let(:array_with_false) { [nil, false] }
+  let(:array_with_true) { [nil, false, true] }
+  describe '#my_each' do
+    it 'returns enum when no block is given' do
+      expect(array.my_each).to be_an Enumerator
+    end
+    it 'returns self when block is given' do
+      expected = array.each { |item| puts item }
+      expect(array.my_each { |item| puts item }).to eql(expected)
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: I don't think the puts are needed here. The only item will cover it. Cause, we are checking it returns self or not.

File: spec/enumerable_spec.rb [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+        expected = array.each_with_index { |item, _index| puts item }
+        expect(array.my_each_with_index { |item, _index| puts item }).to eql(expected)
+      end
+    end
+    context 'when operation is given for item' do
+      it 'loops to each item' do
+        expected = array.each_with_index { |item, _index| item + 1 }
+        expect(array.my_each_with_index { |item, _index| item + 1 }).to eql(expected)
+      end
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: Well, based on the context it is clear to me that you are checking your method on specific criteria. But don't you think that it actually checks returns self or not. You can check it with something like this :point_down: 
```ruby
context 'when operation is given for item' do
  it 'loops to each item' do
    expected = 0
    array.each_with_index { |item, _index| expected += item + 1 }
    calculated = 0
    array.my_each_with_index { |item, _index| calculated += item + 1 } 
    expect(calculated).to eql(expected)
  end
end
```
Then you can try to refactor it later.

File: spec/enumerable_spec.rb [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    context 'when operation is given for index' do
+      it 'loops to each index ' do
+        expected = array.each_with_index { |_item, index| index + 1 }
+        expect(array.my_each_with_index { |_item, index| index + 1 }).to eql(expected)
+      end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: Same as above :point_up:

File: spec/enumerable_spec.rb [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+      it 'loops to each index' do
+        expected = array.each_with_index { |_item, index| index * 1 }
+        expect(array.my_each_with_index { |_item, index| index * 1 }).to eql(expected)
+      end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: I think the context 'loops to each index' seems unstable to me. It would rather say 'iterates each element with respective index'

File: spec/enumerable_spec.rb [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        expect(array.my_select { |item| item % 3 == 0 }).to eql(expected)
+      end
+    end
+  end
+
+  describe '#my_all?' do
+    context 'when no block is given' do
+      it 'items are not false or nil' do
+        expect(array_string.my_all?).to eql(array_string.all?)
+      end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: The context says 'items are not false or nil'. But does not say what it should do. I would rather say. 'returns true if items are not nil or false'.

File: spec/enumerable_spec.rb [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+      it 'items are false or nil' do
+        expect(array_with_nil.my_all?).to eql(array_with_nil.all?)
+      end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: same as above. 'returns false if any of the items is nil or false'.

File: spec/enumerable_spec.rb [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+      it 'items are not given' do
+        expect([].my_all?).to eql([].all?)
+      end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: 'returns true if items are not given'

File: spec/enumerable_spec.rb [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    end
+    context 'when block is given' do
+      it 'items are not false or nil' do
+        expected = array_string.all? { |word| word.length >= 3 }
+        expect(array_string.my_all? { |word| word.length >= 3 }).to eql(expected)
+      end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: it should be like 'if items are not false or nil returns true if the block never returns false.'. Doesn't it sound more concrete?

File: spec/enumerable_spec.rb [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+      it 'items are not false or nil' do
+        expected = array_string.all? { |word| word.length >= 4 }
+        expect(array_string.my_all? { |word| word.length >= 4 }).to eql(expected)
+      end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: You are checking a negative scenario. But is the message defines it? It does not seem like that to me. Maybe it should seem like 'if items are not false or nil returns false when the block returns false.'

File: spec/enumerable_spec.rb [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+    end
+    context 'when regular expression is given' do
+      it 'items are not false or nil' do
+        expected = array_string.all?(/t/)
+        expect(array_string.my_all?(/t/)).to eql(expected)
+      end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: - The context should be 'returns true when elements match with the given expression'

File: spec/enumerable_spec.rb [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+    end
+    context 'when class is given' do
+      it 'items are not false or nil' do
+        expected = array_with_invalid_numeric.all?(Numeric)
+        expect(array_with_invalid_numeric.my_all?(Numeric)).to eql(expected)
+      end
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: - This looks more meaningful, isn't it? 'Items are not nil or false returns true when each item is an instance of the given class'.

File: spec/enumerable_spec.rb [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+  end
+
+  describe '#my_any?' do
+    context 'when no block is given' do
+      it 'returns false' do
+        expect([].my_any?).to eql(false)
+      end
+      it 'items are false or nil' do
+        expect(array_with_nil.my_any?).to eql(array_with_nil.any?)
+      end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: You can change the text the same as above.

File: spec/enumerable_spec.rb [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+    end
+    context 'when block is given' do
+      it 'items are not false or nil' do
+        expected = array_string.any? { |word| word.length >= 3 }
+        expect(array_string.my_any? { |word| word.length >= 3 }).to eql(expected)
+      end
+      it 'items are not false or nil' do
+        expected = array_string.any? { |word| word.length >= 4 }
+        expect(array_string.my_any? { |word| word.length >= 4 }).to eql(expected)
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: Messages should be more clear. So check the provided messages and you can modify it to update for this one.

File: spec/enumerable_spec.rb [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+      end
+    end
+    context 'when regular expression is given' do
+      it 'items are not false or nil' do
+        expected = array_string.any?(/t/)
+        expect(array_string.my_any?(/t/)).to eql(expected)
+      end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: same as above.

File: spec/enumerable_spec.rb [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+    end
+    context 'when class is given' do
+      it 'items are false or nil' do
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: Needs more info.

File: spec/enumerable_spec.rb [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+        expected = array.any?(Numeric)
+        expect(array.my_any?(Numeric)).to eql(expected)
+      end
+      it 'items are not false or nil' do
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: you have to add more info related to the task.

File: spec/enumerable_spec.rb [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+      end
+    end
+  end
+
+  describe '#my_none?' do
+    context 'when no block is given' do
+      it 'returns false' do
+        expect([].my_none?).to eql(true)
+      end
+      it 'items are nil' do
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: If items are nil should return what?

File: spec/enumerable_spec.rb [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+        expect(array_with_one_nil.my_none?).to eql(array_with_one_nil.none?)
+      end
+      it 'items are false or nil' do
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: Add info related to the task covered. Same as above.

File: spec/enumerable_spec.rb [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+        expect(array_with_false.my_none?).to eql(array_with_false.none?)
+      end
+      it 'at least one item is true' do
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: If so what it should return?

File: spec/enumerable_spec.rb [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+        expect(array_with_true.my_none?).to eql(array_with_true.none?)
+      end
+    end
+    context 'when block is given' do
+      it 'items are not false or nil' do
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: Same as above.

File: spec/enumerable_spec.rb [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+        expected = array_string.none? { |word| word.length == 5 }
+        expect(array_string.my_none? { |word| word.length == 5 }).to eql(expected)
+      end
+      it 'items are not false or nil' do
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: same...

File: spec/enumerable_spec.rb [hunk 22]
```diff
@@ -0,0 +1,296 @@
...
+        expected = array_string.none? { |word| word.length >= 4 }
+        expect(array_string.my_none? { |word| word.length >= 4 }).to eql(expected)
+      end
+    end
+    context 'when regular expression is given' do
+      it 'items are not false or nil' do
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: Adds the case what it should return?

File: spec/enumerable_spec.rb [hunk 23]
```diff
@@ -0,0 +1,296 @@
...
+        expected = array_string.none?(/d/)
+        expect(array_string.my_none?(/d/)).to eql(expected)
+      end
+    end
+    context 'when class is given' do
+      it 'items are false or nil' do
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: Decorate the message to render the proper case.

File: spec/enumerable_spec.rb [hunk 24]
```diff
@@ -0,0 +1,296 @@
...
+        expected = array_with_true.none?(Numeric)
+        expect(array_with_true.my_none?(Numeric)).to eql(expected)
+      end
+      it 'items are not false or nil' do
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: same...

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :space_invader: Required Changes :space_invader: 
___
Hi @enelesmai and @rubenpazch :wave: 
Awesome job :clap: . You have added tons of test cases and all of then are executing properly. But have found some issues that are blocking me to approve it right now. But if you feel like these are not an issue, don't hesitate to knock me.
___
### :spades [Mandatory Changes] :spades: 
___
1. We attach messages with each of our tests related to their task. Now the question is why? The short and is, so that we can see what are the corner cases we have covered have added and for which scenario it fails. But some of your messages are not mentioning specifically. Check the comments below. You have to modify them.
2. I have found a test case that is not checking the scenario that it should. So provided an example for it. 
3. How to run tests? You can add the info to your readme. Maybe like `rspec spec/enumerable_spec.rb -fdoc`
___
#### Please fix the above-mentioned issues and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/) | [Twitter](https://twitter.com/Shshamim090)

File: spec/enumerable_spec.rb (hunk 1 above)
Comment (rubenpazch) [reply spec/enumerable_spec.rb]: fixed, thank you for the suggestion

File: spec/enumerable_spec.rb (hunk 2 above)
Comment (rubenpazch) [reply spec/enumerable_spec.rb]: fixed, thank you for the suggestion

File: spec/enumerable_spec.rb (hunk 3 above)
Comment (rubenpazch) [reply spec/enumerable_spec.rb]: fixed, thank you for the suggestion

File: spec/enumerable_spec.rb (hunk 4 above)
Comment (rubenpazch) [reply spec/enumerable_spec.rb]: fixed, thank you for the suggestion

File: spec/enumerable_spec.rb (hunk 5 above)
Comment (enelesmai) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 11 above)
Comment (enelesmai) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 10 above)
Comment (enelesmai) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 9 above)
Comment (enelesmai) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 12 above)
Comment (enelesmai) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 8 above)
Comment (enelesmai) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 7 above)
Comment (enelesmai) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 6 above)
Comment (enelesmai) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 13 above)
Comment (enelesmai) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 14 above)
Comment (enelesmai) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 16 above)
Comment (enelesmai) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 15 above)
Comment (enelesmai) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 17 above)
Comment (rubenpazch) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 18 above)
Comment (rubenpazch) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 19 above)
Comment (rubenpazch) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 20 above)
Comment (rubenpazch) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 21 above)
Comment (rubenpazch) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 22 above)
Comment (rubenpazch) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 23 above)
Comment (rubenpazch) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

File: spec/enumerable_spec.rb (hunk 24 above)
Comment (rubenpazch) [reply spec/enumerable_spec.rb]: Fixed, thanks for the suggestion

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team      
#### Great Job so far, but there are a few more things required. 
### General
- [x] Am getting this error when ever i run the test.
`An error occurred while loading ./spec/enumerable_spec.rb. - Did you mean?`
                   ` rspec ./spec/spec_helper.rb`
`Failure/Error: require 'rspec'`
- [x] Kindly remove this file `rspec_results.html`





 

### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Review by stratospherique — APPROVED** [review]
### Great work implementing the suggested changes :clap: Congrats, the project is approved :heavy_check_mark: You may merge :+1: :+1: :+1: 

Keep up the good work :muscle: 
**_Ahmed Mahfoudh_**
